### ISSUE TITLE
perf: NF4 QLoRA training ~2x slower than PyTorch — Flash Attention + kernel fusion needed

### ISSUE THREAD

**Issue by noahgift** [paiml/entrenar#268]
## Problem

entrenar NF4 QLoRA training is ~50-70% the throughput of PyTorch + bitsandbytes + Flash Attention 2 on the same hardware. Measured on Qwen3-4B:

| Setup | tok/s (RTX 4090) | tok/s (GB10) |
|-------|-----------------|--------------|
| entrenar | ~1,670 | ~17 |
| PyTorch+FA2+bnb (est.) | ~2,500-3,500 | ~30-50 |

### 1. No Flash Attention (P0 — largest single factor)
- entrenar materializes full [S,S] attention matrix: O(S²) memory, O(S²) bandwidth
- Flash Attention 2/3 fuse Q×K→softmax→×V into a single kernel: O(S) memory
- Impact: ~30-50% throughput improvement expected for seq_len=512
- Reference: Dao et al. (2022) "FlashAttention: Fast and Memory-Efficient Exact Attention"

### 2. No CUTLASS/Triton-grade GEMM kernels (P1)
- trueno generates PTX by string templating with fixed tile sizes
- CUTLASS uses compile-time tile scheduling optimized per GPU arch
- Impact: ~10-20% on compute-bound layers

### 3. No kernel fusion (P1)
- Separate kernel launches for RMSNorm, attention, FFN sub-ops
- PyTorch/Triton fuse sequences like RMSNorm+QKV projection
- Impact: ~5-15% from reduced launch overhead and memory traffic

### 4. JIT compilation overhead on new architectures (P2)
- sm_70 PTX retargeting works but produces suboptimal SASS
- Native sm_121 kernels would be faster on Blackwell
- Impact: unknown, possibly 5-10%

## Proposed Fix: Flash Attention CUDA Kernel

Implement FlashAttention-2 algorithm in trueno-gpu:

```
// Tiled attention: for each block of Q rows
for q_block in Q.chunks(Br):
    // Online softmax accumulation
    for k_block in K.chunks(Bc):
        S = q_block @ k_block.T / sqrt(d)
        m_new = max(m_old, rowmax(S))
        P = exp(S - m_new)
        O = diag(exp(m_old - m_new)) @ O_old + P @ v_block
        l = exp(m_old - m_new) * l_old + rowsum(P)
    O = O / l
```

Key design decisions:
- PTX kernel with shared memory tiling (Br=64, Bc=64 for sm_70+)
- Online softmax (no materialized [S,S] matrix)
- Backward: recompute attention in chunks (no activation storage)
- GQA support (num_kv_heads < num_heads) for Qwen3-4B

## Success Criteria

- [ ] Flash Attention forward kernel passes numerical tests (vs naive attention, atol=1e-3)
- [ ] Flash Attention backward kernel produces correct gradients
- [ ] >25% throughput improvement on Qwen3-4B QLoRA training
- [ ] Memory usage drops (no [S,S] materialization)
- [ ] Works on both Ada (sm_89) and Blackwell (sm_121)

## References

- [FlashAttention-2](https://arxiv.org/abs/2307.08691) — Dao (2023)
- [FlashAttention-3](https://arxiv.org/abs/2407.08691) — Shah et al. (2024), Hopper/Blackwell specific
- [CUTLASS Flash Attention](https://github.com/NVIDIA/cutlass/tree/main/examples/41_fused_multi_head_attention)
